### ISSUE TITLE
[memory leak] [PyTorch] create_graph=True w/ custom grad 

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#7343]
```python
>>> import torch
>>> import gc
>>> _ = torch.randn(1, device='cuda')
>>> del _
>>> torch.cuda.synchronize()
>>> gc.collect()
0
>>> print(torch.cuda.memory_allocated())
865280
>>> x = torch.randn(1, device='cuda', requires_grad=True)
>>> y = x.tanh()
>>> y.backward(torch.ones_like(y), create_graph=True)
>>> del x, y
>>> torch.cuda.synchronize()
>>> gc.collect()
0
>>> print(torch.cuda.memory_allocated())
867328
```

leaks with `y  = x.tanh()` but not `y = x  +1`.

Discovered when running code in #7270

**Comment by apaszke** [comment]
That's a known issue, that's hard to solve in general. The problem is that we get a reference cycle in C++, which used to be broken by having a `weak_ptr` in the `AccumulateGrad`, which is now a strong reference. One way to get around it would be to disallow `create_graph` with `.backward` (and only use it with `.grad`).

IIRC the reference cycle is `x -> x.grad -> x.grad.grad_fn = TanhBackwardBackward -> x`

**Comment by colesbury** [comment]
@apaszke just beat me to describing the issue. To expand on his point, I think we need to disable history on `Tensor.grad`. So you would need to use `torch.autograd.grad` instead of `Tensor.backward()`

**Comment by ezyang** [comment]
related https://github.com/pytorch/pytorch/pull/24368

**Comment by albanD** [comment]
Note that since then, we added a couple warnings to the doc to try and deter users from doing this.
Also note that people that use the new `zero_grad(set_to_none=True)` will break the cycle and avoid the leak !

**Comment by motiwari** [comment]
@albanD 

> Note that since then, we added a couple warnings to the doc to try and deter users from doing this.
> Also note that people that use the new `zero_grad(set_to_none=True)` will break the cycle and avoid the leak !

@albanD this fix doesn't work for me -- do you know why this might be the case?

**Comment by albanD** [comment]
Can you open a topic on the forum https://discuss.pytorch.org/ with details of your training loop and what you do?

**Comment by hweekuan** [comment]
I am facing the same problem. even with del and autograd.grad, create_graph=True gives memory leak. I distill the code here

    torch.manual_seed(2332)
    dim = 10
    winit = torch.randn([dim])
    w1 = torch.randn([dim+1],requires_grad=True)
    w2 = w1[dim:]
    torch.autograd.set_detect_anomaly(True)

    for e in range(1000):
        param0 = winit.clone().detach().requires_grad_()
    
        p = torch.sum((param0-w2)*(param0-w2))
        param0grad = torch.autograd.grad(p,param0,create_graph=True)[0] #mem leak
        param1 = param0 - 0.1*param0grad

        del param1
        del param0grad
        del p
        del param0
        gc.collect()

        if (e%100==0): 
            p = psutil.Process()
            mem_info = p.memory_info()
            print('epoch ',e,' mem info ',mem_info)

========
run time output:
epoch  0  mem info  pmem(rss=95322112, vms=4828065792, pfaults=24924, pageins=135)
epoch  100  mem info  pmem(rss=95776768, vms=4828065792, pfaults=25040, pageins=135)
epoch  200  mem info  pmem(rss=96342016, vms=4830425088, pfaults=25178, pageins=135)
epoch  300  mem info  pmem(rss=96907264, vms=4830687232, pfaults=25316, pageins=135)
epoch  400  mem info  pmem(rss=97456128, vms=4831997952, pfaults=25450, pageins=135)
epoch  500  mem info  pmem(rss=97988608, vms=4832260096, pfaults=25580, pageins=135)

thank you.

**Comment by ptrblck** [comment]
@hweekuan you are using anomaly detection via `torch.autograd.set_detect_anomaly(True)`, which is a debug tool and was causing an increasing memory usage in the past.
Could you disable it, once you've debugged your code (this mode would also slow down your code so should only be used while debugging)?

**Comment by albanD** [comment]
Which version of pytorch are you using?
Running your script on colab doesn't show any leak:
```

Colaboratory logo

What is Colaboratory?
Colaboratory, or "Colab" for short, allows you to write and execute Python in your browser, with

Zero configuration required
Free access to GPUs
Easy sharing
Whether you're a student, a data scientist or an AI researcher, Colab can make your work easier. Watch Introduction to Colab to learn more, or just get started below!

Getting started
The document you are reading is not a static web page, but an interactive environment called a Colab notebook that lets you write and execute code.

For example, here is a code cell with a short Python script that computes a value, stores it in a variable, and prints the result:

[4]
1m
import torch
import psutil
import gc

torch.manual_seed(2332)
dim = 10
winit = torch.randn([dim])
w1 = torch.randn([dim+1],requires_grad=True)
w2 = w1[dim:]
torch.autograd.set_detect_anomaly(True)

epoch  0  mem info  pmem(rss=283746304, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  100  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  200  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  300  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  400  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  500  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  600  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  700  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  800  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
epoch  900  mem info  pmem(rss=283979776, vms=2241384448, shared=137732096, text=4620288, lib=0, data=461254656, dirty=0)
```

It is possible that your malloc library is caching some memory and not returning it to the OS as there is no memory pressure.

**Comment by hweekuan** [comment]
@albanD thank you, my versions: Python 3.7.6, >>> torch.__version__ '1.7.1'

@ptrblck you are right, removed this line and no more memory leak
#torch.autograd.set_detect_anomaly(True)

thank you so much. I will post the answer in the other group to help others.
